File: VB/Extensions/Chart/ChartField.vb
Imports DevExpress.Office.Utils
Imports DevExpress.XtraRichEdit.Model
Imports DevExpress.XtraRichEdit.Fields
Imports System.Collections.Generic

Namespace BizPad

    Public Class ChartField
        Inherits CalculatedFieldBase

        Public Shared ReadOnly FieldType As String = "CHART"
        Protected Overrides ReadOnly Property FieldTypeName() As String
            Get
                Return FieldType
            End Get
        End Property

        Private Shared ReadOnly switchesWithArgument As Dictionary(Of String, Boolean) = CreateSwitchesWithArgument("w", "h", "d", "c", "s")
        Protected Overrides ReadOnly Property SwitchesWithArguments() As Dictionary(Of String, Boolean)
            Get
                Return switchesWithArgument
            End Get
        End Property

        Private chart As New ChartImage()

        Public Overrides Sub Initialize(ByVal pieceTable As PieceTable, ByVal switches As InstructionCollection)
            MyBase.Initialize(pieceTable, switches)
            chart.Initialize(switches)
        End Sub


        Public Overrides Function GetCalculatedValueCore(ByVal sourcePieceTable As PieceTable, ByVal mailMergeDataMode As MailMergeDataMode, ByVal documentField As Field) As CalculatedFieldValue

            Dim image As OfficeImage = chart.CreateImage()
            Dim targetModel As DocumentModel = sourcePieceTable.DocumentModel.GetFieldResultModel()
            targetModel.MainPieceTable.InsertInlinePicture(DocumentLogPosition.Zero, image)
            Return New CalculatedFieldValue(targetModel)
        End Function
    End Class
End Namespace

File: CS/Extensions/Chart/ChartField.cs
namespace BizPad {
    using System.Collections.Generic;
    using DevExpress.XtraRichEdit.Fields;
    using DevExpress.XtraRichEdit.Model;
    using DevExpress.XtraRichEdit.Utils;

    public class ChartField : CalculatedFieldBase {
        public static readonly string FieldType = "CHART";
        protected override string FieldTypeName {
            get {
                return FieldType;
            }
        }

        static readonly Dictionary<string, bool> switchesWithArgument = CreateSwitchesWithArgument("w", "h", "d", "c", "s");
        protected override Dictionary<string, bool> SwitchesWithArguments {
            get {
                return switchesWithArgument;
            }
        }

        ChartImage chart = new ChartImage();

        public override void Initialize(InstructionCollection switches) {
            base.Initialize(switches);
            chart.Initialize(switches);
        }


        public override CalculatedFieldValue GetCalculatedValueCore(PieceTable sourcePieceTable,
            MailMergeDataMode mailMergeDataMode,
            Field documentField) {

            RichEditImage image = chart.CreateImage();
            DocumentModel targetModel = sourcePieceTable.DocumentModel.GetFieldResultModel();
            targetModel.MainPieceTable.InsertInlinePicture(DocumentLogPosition.Zero, image);
            return new CalculatedFieldValue(targetModel);
        }
    }
}
